### DISCUSSION TITLE
Typescript doesn't recognize prisma return types

### DISCUSSION THREAD

**Discussion by Bradleykingz** [prisma/prisma#3253] [Q&A] [answered]
Typescript keeps bugging out and throwing the following error

`Type '{}' is missing the following properties from type 'AccountType': email, password`

Where `AccountType` is a type I've created that extends prisma's generated output.

```ts
import {Account} from "@prisma/client"
export interface AccountType extends Account {
  id?: number;
  username?: string;
}
```

And `Account` from prisma's generated output looks like:

```ts
export type Account = {
  email: string
  password: string
} 
```
Basically, "email" and "password" are required, while "username" and "id" are optional.

For whatever reason, Typescript seems to think this function returns an empty object

```ts
  async findOneWhere(
    where: AccountWhereUniqueInput,
    select?: AccountSelect,
    include?: AccountInclude
  ): Promise<AccountType> {
    // No need to look further.
    return prisma.account.findOne({
      where,
      select,
    });
  }
```
I'm not sure why it happens, but this is the third time I've had to git clone my project afresh because I have no clue how to even begin debugging it :shrug:

Any pointers would be appreciated. 

**Edit:** FWIW, this only happens with `select`. With `include` it works fine.

**Comment by ryands17** [comment]
Hey @Bradleykingz 👋 
Could you share your Account model so that I can check from my end?

**Reply by Bradleykingz** [reply]
It looks like this

```prisma
model Account {
  id                   Int           @default(autoincrement()) @id
  username             String
  email                String
  password             String
  randomPassword       String?
}
```
But I don't think it's the model that's a problem. I've encountered it pretty randomly each time, and it affects all my models.

**Reply by ryands17** [reply, marked as answer]
This is again due to a TS issue. You are dynamically providing select, so TS cannot infer the arguments at runtime. Event if I do not extend any interface, I will face this issue as you can see here:

```ts
import * as Prisma from '@prisma/client'

const prisma = new Prisma.PrismaClient()

async function findOneWhere(
  where: Prisma.AccountWhereUniqueInput,
  select?: Prisma.AccountSelect
) {
  return prisma.account.findOne({
    where,
    select,
  })
}

// calling the function
findOneWhere(
  { id: 2 },
  {
    id: true,
    email: true,
  }
)
  .then((res) => {
    console.log(res.id)
  })
```

![image](https://user-images.githubusercontent.com/19697099/89776992-897bfb00-db28-11ea-83a6-eeeef7b071a6.png)

As you can see here it cannot dynamically fetch the types.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
